File: petclinic-backend-ts/src/pets/dto/pet-fields.dto.ts
import { ApiProperty } from '@nestjs/swagger';
import { Type } from 'class-transformer';
import { IsDefined, MaxLength, ValidateNested } from 'class-validator';
import { PetTypeDto } from '../../pet-types/dto/pet-type.dto';
import { IsPastOrPresent } from '../../common/validation/is-past-or-present.validator';

/**
 * Mutable fields accepted when creating or updating a pet.
 *
 * birthDate is an ISO 'YYYY-MM-DD' string.
 */
export class PetFieldsDto {
  @IsDefined()
  @MaxLength(30)
  @ApiProperty({ example: 'Leo', description: 'The name of the pet.' })
  name!: string;

  @IsDefined()
  @IsPastOrPresent({ message: 'Birth date must not be in the future' })
  @ApiProperty({ example: '2010-09-07', description: 'The date of birth of the pet.' })
  birthDate!: string;

  @IsDefined()
  @ValidateNested()
  @Type(() => PetTypeDto)
  @ApiProperty({ type: () => PetTypeDto })
  type!: PetTypeDto;
}

File: petclinic-backend/src/main/java/victor/training/petclinic/rest/dto/PetFieldsDto.java
package victor.training.petclinic.rest.dto;

import java.time.LocalDate;
import org.springframework.format.annotation.DateTimeFormat;
import jakarta.validation.Valid;
import jakarta.validation.constraints.*;
import io.swagger.v3.oas.annotations.media.Schema;

public class PetFieldsDto {

    @NotNull
    @Size(max = 30)
    @Schema(example = "Leo", description = "The name of the pet.")
    private String name;

    @DateTimeFormat(iso = DateTimeFormat.ISO.DATE)
    @NotNull
    @Valid
    @Schema(example = "2010-09-07", description = "The date of birth of the pet.")
    @PastOrPresent(message = "Birth date must not be in the future")
    private LocalDate birthDate;

    @NotNull
    @Valid
    @Schema
    private PetTypeDto type;

    public String getName() {
        return name;
    }

    public void setName(String name) {
        this.name = name;
    }

    public LocalDate getBirthDate() {
        return birthDate;
    }

    public void setBirthDate(LocalDate birthDate) {
        this.birthDate = birthDate;
    }

    public PetTypeDto getType() {
        return type;
    }

    public void setType(PetTypeDto type) {
        this.type = type;
    }
}
